pkg/zson/raw.go

```diff
@@ -220,3 +223,132 @@ func zsonParseField(b []byte) ([]byte, []byte, error) {
+}
+
+func jsonParseObject(b []byte) ([][]byte, zeek.Type, error) {
+	type kv struct{ column string }
+	var columns []zeek.Column
+	var values [][]byte
+	err := jsonparser.ObjectEach(b, func(key []byte, value []byte, typ jsonparser.ValueType, offset int) error {
+		val, vtyp, err := jsonParseValue(value, typ)
+		if err != nil {
+			return err
+		}
+		values = append(values, val)
+		columns = append(columns, zeek.Column{Name: string(key), Type: vtyp})
+		return nil
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	// sort fields lexigraphically ensuring maps with the same
+	// columns but different printed order get assigned the same descriptor.
+	sort.Slice(columns, func(i, j int) bool {
+		v := columns[i].Name < columns[j].Name
+		if v {
+			values[i], values[j] = values[j], values[i]
+		}
+		return v
+	})
+	return values, &zeek.TypeRecord{Columns: columns}, nil
+}
+
+func jsonParseValue(raw []byte, typ jsonparser.ValueType) ([]byte, zeek.Type, error) {
+	switch typ {
+	default:
+		return nil, nil, fmt.Errorf("unsupported type %v", typ)
+	case jsonparser.Array:
+		return jsonParseArray(raw)
+	case jsonparser.Object:
+		values, typ, err := jsonParseObject(raw)
+		if err != nil {
+			return nil, nil, err
+		}
+		return zval.AppendContainer(nil, values), typ, err
+	case jsonparser.Boolean:
+		return jsonParseBool(raw)
+	case jsonparser.Number:
+		return jsonParseNumber(raw)
+	case jsonparser.Null:
+		return jsonParseString(nil)
+	case jsonparser.String:
+		return jsonParseString(raw)
+	}
+}
+
+func jsonParseBool(raw []byte) ([]byte, zeek.Type, error) {
+	b, err := jsonparser.GetBoolean(raw)
+	if err != nil {
+		return nil, nil, err
+	}
+	value := strconv.AppendBool(nil, b)
+	return value, zeek.TypeBool, err
+}
+
+// XXX This needs to handle scientific notation... I think.
+func jsonParseNumber(b []byte) ([]byte, zeek.Type, error) {
+	if idx := bytes.IndexRune(b, '.'); idx == -1 {
+		return b, zeek.TypeInt, nil
+	}
+	return b, zeek.TypeDouble, nil
+}
+
+// XXX do escaping guff
+func jsonParseString(b []byte) ([]byte, zeek.Type, error) {
+	return b, zeek.TypeString, nil
+}
+
+func jsonParseArray(raw []byte) ([]byte, zeek.Type, error) {
+	var err error
+	var values [][]byte
+	var types []zeek.Type
+	jsonparser.ArrayEach(raw, func(el []byte, typ jsonparser.ValueType, offset int, elErr error) {
+		if err != nil {
+			return
+		}
+		if elErr != nil {
+			err = elErr
+		}
+		var val []byte
+		var ztyp zeek.Type
+		val, ztyp, err = jsonParseValue(el, typ)
+		values = append(values, val)
+		types = append(types, ztyp)
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	var vType zeek.Type
+	for _, t := range types {
+		if vType == nil {
+			vType = t
+		} else if vType != t {
+			vType = nil
+			break
+		}
+	}
+	// XXX zeek currently doesn't support sets with heterogeneous types so for
+	// now just convert values to strings
+	if vType == nil {
+		vType = zeek.TypeString
+	}
+	return zval.AppendContainer(nil, values), &zeek.TypeVector{Type: vType}, nil
```

Comment:
Creating a zeek.TypeVector like this breaks zeek.SameType, so you should use zeek.LookupType or create a new function in the zeek package along the lines of `LookupVectorType(element Type) Type`.

Reply:
Great catch, thanks.